src/base/dev/vga/vgaemu.c

```diff
@@ -805,15 +803,19 @@ unsigned vga_read_dword(dosaddr_t addr)
 void vga_mark_dirty(dosaddr_t s_addr, int len)
 {
   unsigned vga_page, abeg, aend, addr;
+  int i, j;
   abeg = s_addr & PAGE_MASK;
   aend = (s_addr + len - 1) & PAGE_MASK;
   for (addr = abeg; addr <= aend; addr += PAGE_SIZE) {
-    if (!vga_write_access(addr))
-      continue;
-    vga_page = (addr >> PAGE_SHIFT) -
-	vga.mem.map[VGAEMU_MAP_BANK_MODE].base_page +
-	vga.mem.map[VGAEMU_MAP_BANK_MODE].first_page;
-    _vga_emu_adjust_protection(vga_page, 0, VGA_PROT_RW, 1);
+    for(i = 0; i < VGAEMU_MAX_MAPPINGS; i++) {
+      j = (addr >> PAGE_SHIFT) - vga.mem.map[i].base_page;
+      if(j >= 0 && j < vga.mem.map[i].pages) {
```

Comment:
Is any check needed here, that
the protection is already RW, so
no need to adjust protection?

Reply:
`_vga_emu_adjust_protection` has that check already (via the `vgaemu_prot_ok` call in `vga_emu_protect_page`). Adding a `vgaemu_prot_ok` here is possible too but that's just efficiency, not necessity, and the efficiency is already taken care of in write_byte and co via the unprotected page cache.

Reply:
But I don't see a single `return` statement
in `_vga_emu_adjust_protection()` that could
return prematurely and avoid doing many things.
What am I missing?

Reply:
`_vga_emu_adjust_protection()` calls `vga_emu_protect()` a number of times depending on the mode, which in turn calls `vga_emu_protect_page()`, which in turn has the early return.